From 45cf411b077e82419f6c2b04ce97043f144071fc Mon Sep 17 00:00:00 2001
From: Melissa Wen <mwen@igalia.com>
Date: Fri, 16 Jan 2026 13:26:01 -0300
Subject: [PATCH 033/188] drm/amd/display: fix screen blinking when exits SubVP
 to MPO
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Whenever transitioning from SubVP to MPO in a high refresh rate and high
resolution display, for example, 4k@120Hz or higher, screen blinks as a
white flash. On gamescope, it happens every time the cursor disappears
because it stops composing in a single plane and returns using primary +
one/two overlay planes. According to [1] SubVP + MPO isn't supported and
the white flash seems to occur exactly when exiting SubVP.

It happens with DCN321 and DCN401, at least. As there is no transition
from SubVP to MPO if SubVP is disabled (ofc), there is no blink when
moving from single plane to multiple planes. Check [2] for more details.

[1] https://gitlab.freedesktop.org/agd5f/linux/-/commit/14eb72ff580d
[2] https://gitlab.freedesktop.org/drm/amd/-/issues/4877

Signed-off-by: Melissa Wen <mwen@igalia.com>
Link: https://lore.kernel.org/amd-gfx/20260116162741.89743-1-mwen@igalia.com/
Signed-off-by: Nícolas F. R. A. Prado <nfraprado@collabora.com>
---
 drivers/gpu/drm/amd/display/dc/resource/dcn32/dcn32_resource.c  | 2 +-
 .../gpu/drm/amd/display/dc/resource/dcn321/dcn321_resource.c    | 2 +-
 .../gpu/drm/amd/display/dc/resource/dcn401/dcn401_resource.c    | 2 +-
 3 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/drm/amd/display/dc/resource/dcn32/dcn32_resource.c b/drivers/gpu/drm/amd/display/dc/resource/dcn32/dcn32_resource.c
index a11110e30..491d3af0b 100644
--- a/drivers/gpu/drm/amd/display/dc/resource/dcn32/dcn32_resource.c
+++ b/drivers/gpu/drm/amd/display/dc/resource/dcn32/dcn32_resource.c
@@ -772,7 +772,7 @@ static const struct dc_debug_options debug_defaults_drv = {
 		}
 	},
 	.use_max_lb = true,
-	.force_disable_subvp = false,
+	.force_disable_subvp = true,
 	.exit_idle_opt_for_cursor_updates = true,
 	.using_dml2 = false,
 	.using_dml21 = false, // TODO : Temporary for N-1 validation. Remove after N-1 is done.
diff --git a/drivers/gpu/drm/amd/display/dc/resource/dcn321/dcn321_resource.c b/drivers/gpu/drm/amd/display/dc/resource/dcn321/dcn321_resource.c
index d1dbcc8dd..0d5016cfd 100644
--- a/drivers/gpu/drm/amd/display/dc/resource/dcn321/dcn321_resource.c
+++ b/drivers/gpu/drm/amd/display/dc/resource/dcn321/dcn321_resource.c
@@ -762,7 +762,7 @@ static const struct dc_debug_options debug_defaults_drv = {
 		}
 	},
 	.use_max_lb = true,
-	.force_disable_subvp = false,
+	.force_disable_subvp = true,
 	.exit_idle_opt_for_cursor_updates = true,
 	.enable_single_display_2to1_odm_policy = true,
 
diff --git a/drivers/gpu/drm/amd/display/dc/resource/dcn401/dcn401_resource.c b/drivers/gpu/drm/amd/display/dc/resource/dcn401/dcn401_resource.c
index e3ee792fb..e715dd28c 100644
--- a/drivers/gpu/drm/amd/display/dc/resource/dcn401/dcn401_resource.c
+++ b/drivers/gpu/drm/amd/display/dc/resource/dcn401/dcn401_resource.c
@@ -760,7 +760,7 @@ static const struct dc_debug_options debug_defaults_drv = {
 		}
 	},
 	.use_max_lb = true,
-	.force_disable_subvp = false,
+	.force_disable_subvp = true,
 	.disable_force_pstate_allow_on_hw_release = false,
 	.exit_idle_opt_for_cursor_updates = true,
 	.using_dml2 = true,
-- 
2.55.0

